src/test/util/setup_common.cpp

```diff
@@ -120,18 +119,49 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", "Custom data directory (default: <random_string>)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    {
+        std::string testdatadir;
+        if (!m_node.args->IsArgSet("-testdatadir")) {
+            testdatadir = g_insecure_rand_ctx_temp_path.rand256().ToString().c_str();
+        } else {
+            testdatadir = m_node.args->GetArg("-testdatadir").value();
+            // Don't allow anything but a simple string (a single path component) because we're
+            // going to remove this path, and it would be bad if the string was "../../home".
+            if (testdatadir.find(fs::path::preferred_separator) != std::string::npos) {
+                std::cout << "testdatadir cannot contain path separators: " << testdatadir << std::endl;
+                exit(1);
+            }
```

Comment:
In 5dd89adbc:
Can drop this now. Custom test data directories are not deleted anymore.

This is good for running test on a different storage location. Not only on the OS temporary files directory.

Reply:
We do remove the custom directory at the beginning of the test:
https://github.com/bitcoin/bitcoin/pull/26564/files#diff-6a8ef76c60f30a6ca67d9f0e478fd02989c4b7fbc4c3116f80e13d873d5775e6R164
so I think this restriction is still needed. I just verified that `/tmp/test_common_Bitcoin\ Core` can be a symlink, so that would allow you to use any location, right? (I could add that to the README but not sure it's worth it.)

Reply:
> We do remove the custom directory at the beginning of the test:
> https://github.com/bitcoin/bitcoin/pull/26564/files#diff-6a8ef76c60f30a6ca67d9f0e478fd02989c4b7fbc4c3116f80e13d873d5775e6R164
> so I think this restriction is still needed. I just verified that `/tmp/test_common_Bitcoin\ Core` can be a symlink, so that would allow you to use any location, right? (I could add that to the README but not sure it's worth it.)

hmm, that wasn't there before. Why we should introduce it?.

We have two cases, default and custom dir path.
The default test behavior creates a random directory path within the OS temp directory; therefore, we don't need to remove anything on this execution path.
Regarding the custom path functionality, I'm not entirely certain we should assign core the responsibility of cleaning up the custom directory before executing the test/s. I think we should just try to lock the folder and run the test, similar to what we do for bitcoind.
Running tests on different paths, not on the temp directory, would be helpful. Doing it has no risk if we don't remove the directory after running the test there. And I think that users who make use of the custom path functionality should be aware of what they are doing

Reply:
> The default test behavior creates a random directory path within the OS temp directory; therefore, we don't need to remove anything on this execution path.

No, in current master, the test framework does delete the directory _after_ the test: https://github.com/bitcoin/bitcoin/blob/3e691258d8789a4a89cce42e7e71b130491594d7/src/test/util/setup_common.cpp#L162 and I think that's a good thing or else running many tests could fill up your temp directory. If the test stops unexpectedly, I think the directory isn't removed, which is probably a good thing so that the cause of the failure can be investigated.

The directory doesn't need to be removed before the test; we can presume the directory doesn't exist because of the random name. (This PR does attempt to remove the random directory, which does no harm but it's not necessary.)

> Regarding the custom path functionality, I'm not entirely certain we should assign core the responsibility of cleaning up the custom directory before executing the test/s.

No, it does need to, because if a (populated) directory exists when the test begins, the test will likely fail because of this unexpected leftover state.

Reply:
I was talking about what we do before the test. The test cleanup after execution is fine for the default case.

I think we shouldn't block the test creation on different paths. Be forced to place the test only inside the temp directory doesn't seems so useful to me.
What if we make the `-testdatadir` arg the base path for all test cases, and create a subdirectory for each run test inside it?. (with this, you would not need to delete the datadir at the beginning, because it will not exist).
E.g.
```c++
fs::path base_dir;
if (!m_node.args->IsArgSet("-testdatadir")) {
    base_dir = fs::temp_directory_path();
} else {
    std::string testdatadir = m_node.args->GetArg("-testdatadir").value();
    // Allow only a simple string (a single path component) because we're
    // going to remove this path, and it would be bad if it was "../../home".
    if (testdatadir.find(fs::path::preferred_separator) != std::string::npos) {
        std::cerr << "-testdatadir cannot contain path separators: " << testdatadir << std::endl;
        exit(1);
    }

    // Allow only regular directory paths
    if (fs::is_regular_file(testdatadir) || fs::is_symlink(testdatadir)) {
        std::cerr << "-testdatadir needs to be a directory path: " << testdatadir << std::endl;
        exit(1);
    }

    // Create base dir if needed
    fs::create_directories(testdatadir);
    base_dir = fs::path{testdatadir.c_str()};
    m_has_custom_datadir = true;
}

// Create test dir within the base directory
m_path_root = base_dir / "test_common_" PACKAGE_NAME / g_insecure_rand_ctx_temp_path.rand256().ToString().c_str();
```

Reply:
This defeats the purpose of this pull because it generates a randomly named data directory path; the point of this PR is to have a fixed datadir pathname so that I can have various files open in an editor across multiple test runs (not have to determine files' locations and re-open them on each test run).

> Be forced to place the test only inside the temp directory doesn't seems so useful to me.

Well, if we want the datadir pathnames to be static (which is what _I'm_ trying to accomplish with this PR), then we _must_ delete the datadir before the test begins. The first version of this PR did indeed allow the datadir to be anywhere, but this comment: https://github.com/bitcoin/bitcoin/pull/26564#issuecomment-1328190540 points out that it's dangerous to recursively remove a directory (because of the possibility of user error) -- but within `/tmp`, and especially within `/tmp/test_common_Bitcoin\ Core/`, recursive directory remove is okay, not dangerous at all.

And, again, the user can set up a symlink from `/tmp/test_common_Bitcoin\ Core/` to anywhere, if there's a desire not to use `/tmp` for some reason (but that should be rare). That reintroduces the danger, of course, but that's why we won't even document setting up a symlink.

Reply:
> And, again, the user can set up a symlink from /tmp/test_common_Bitcoin\ Core/ to anywhere, if there's a desire not to use /tmp for some reason (but that should be rare).

Have you seen https://github.com/bitcoin/bitcoin/pull/26684#pullrequestreview-1566578747 and https://github.com/bitcoin/bitcoin/pull/28955#issuecomment-1829948567?
It seems that we need it.

> Well, if we want the datadir pathnames to be static (which is what I'm trying to accomplish with this PR), then we must delete the datadir before the test begins. The first version of this PR did indeed allow the datadir to be anywhere, but this comment: https://github.com/bitcoin/bitcoin/pull/26564#issuecomment-1328190540 points out that it's dangerous to recursively remove a directory (because of the possibility of user error) -- but within /tmp, and especially within /tmp/test_common_Bitcoin\ Core/, recursive directory remove is okay, not dangerous at all.

Yes. I understood why you added and the history. What I proposed was a middle ground between the "only run tests in a /temp/<something> directory" option and the "run test on any custom path" option. It was an option to allow the user to provide the base directory path where the test will be created and run without being deleted after it execution.

Other than that, I'm just not sure why we need to delete the directory path recursively at the beginning of the test. Why we cannot delete only what the test will use? or.. don't delete anything and move such responsibility to the user.
--- Also, I have to admit that I'm not a fan of the symlink approach but.. thats a different topic ---

Reply:
I hadn't seen those PRs, thanks. I tried running `test_bitcoin` with the `TMPDIR` environment variable set (on Linux), and it works, so that means you don't have to set up a symlink.

> or.. don't delete anything and move such responsibility to the user.

Yes, I like that suggestion. I had forgotten that vscode (which is where I run the debugger from) allows one to set up an arbitrary "task" to run, `preLaunchTask`, before starting the executable, and this can remove the (fixed-named) data directory, so that's perfect. Much less dangerous than recursive remove.

So what do you think about this:

- change the argument name to `-datadir=` instead of `-testdatadir=` (since the argument will have the same meaning as for other executables like `bitcoind`)
- if `-datadir` is not specified, behave as today
- have the `datadir` argument be the actual full pathname to use for the data directory (can have path separators, and be a relative name or absolute path, we don't care)
- _don't_ delete this directory before starting the test (could possibly warn if it's non-empty? it really should be empty when we start)
- if `-datadir=` is specified, don't delete the directory after the test completes.

Reply:
Sounds good to me.
Probably the only reason we may want to abort running a test on a specific datadir is when the directory contain mainnet data. But it shouldn't be a problem as we should be having something to detect this scenario inside `init.cpp` already.

Reply:
That's an excellent point, but I doubt there is anything already present that will prevent this.

Here are two ideas, what do you think?

- Whatever argument is given to `-datadir=`, _always_ make (if necessary) and use a subdirectory (within the specified directory) called `test_bitcoin`. That way, even if the user specifies their real data directory, all the test stuff will be within this subdirectory. (It's similar to how there already are subdirectories for testnet, signet, and regtest.)
- make the argument keyword `-testdatadir=` (like it is now) instead of `-datadir=` (which might lead someone to think of specifying the real data directory path).

The first option is probably safer. Actually, I might do both, because this `-datadir` argument would behave a little differently (by making that subdirectory) than what people may be expecting. Also, there's a technical reason why `-datadir=` is a little more difficult to do (I won't bother to explain it right here).

If I have a chance, I'll see if you're correct that it's already prevented.

Reply:
> Whatever argument is given to -datadir=, always make (if necessary) and use a subdirectory (within the specified directory) called test_bitcoin. That way, even if the user specifies their real data directory, all the test stuff will be within this subdirectory. (It's similar to how there already are subdirectories for testnet, signet, and regtest.)

Sounds good. s/test_bitcoin/tests (we don't use "signet_bitcoin", "testnet_bitcoin", "regtest_bitcoin").

------------

Also, probably something for another PR but.. the following feature came to my mind: https://github.com/furszy/bitcoin-core/commits/2023_test_nocleanup/. Check my two commits there.

Reply:
Thanks for the suggestion; I decided to name the extra path component `test_temp` instead of `tests`, because if the user thought it was okay to specify their home directory, they may have a directory there called `tests` with some source code. This name isn't likely to conflict, and the `temp` part of the name indicates that it's a temporary directory (unlike `signet`, `regtest`, etc., which make sense to persist).